core/amber/src/main/python/core/models/RTupleExecutor.py

```diff
@@ -0,0 +1,193 @@
+            binary_r_list: dict[str, object] = {}
+            for k, v in binary_tuple.as_dict().items():
+                if isinstance(v, bytes):
+                    binary_r_list[k] = pickle.loads(v[10:])
+                elif isinstance(v, datetime.datetime):
+                    binary_r_list[k] = robjects.vectors.POSIXct.sexp_from_datetime([v])
+                else:
+                    binary_r_list[k] = v
+
+            binary_r_list: rpy2.robjects.ListVector = robjects.vectors.ListVector(
+                binary_r_list
+            )
+
+            input_r_list: rpy2.robjects.ListVector = (
+                RTupleExecutor._combine_binary_and_non_binary_lists(
+                    non_binary_pyarrow_array, binary_r_list
+                )
+            )
+
+            output_r_generator: rpy2.robjects.SignatureTranslatedFunction = self._func(
+                input_r_list, port
+            )
+
+            while True:
+                output_r_tuple: rpy2.robjects.ListVector = output_r_generator()
+                if (
+                    isinstance(output_r_tuple, rinterface.SexpSymbol)
+                    and str(output_r_tuple) == ".__exhausted__."
+                ):
+                    break
+                if isinstance(output_r_tuple.names, rpy2.rinterface_lib.sexp.NULLType):
+                    yield None
+                    break
+
+                diff_fields: list[str] = [
+                    field_name
+                    for field_name in output_r_tuple.names
+                    if field_name not in input_fields
+                ]
+
+                output_python_dict: dict[str, object] = {
+                    key: output_r_tuple.rx2(key) for key in (input_fields + diff_fields)
+                }
+                output_python_dict: dict[str, object] = {
+                    key: _convert_r_to_py(value)
+                    for key, value in output_python_dict.items()
+                }
+                yield Tuple(output_python_dict)
+
+
+class RTupleSourceExecutor(SourceOperator):
+    """
+    A source operator that produces a generator that yields R Lists using R code.
+    R Lists are R's representation of a Tuple
+    """
+
+    is_source = True
+
+    def __init__(self, r_code: str):
+        """
+        Initialize the RTupleSourceExecutor with R code.
+
+        Args:
+            r_code (str): R code to be executed.
+        """
+        super().__init__()
+        # Use the local converter from rpy2 to load in the R function given by the user
+        with local_converter(default_converter):
+            self._func = robjects.r(r_code)
+
+    def produce(self) -> Iterator[Union[TupleLike, TableLike, None]]:
+        """
+        Produce Tuples using the provided R function.
+        Used by the source operator only.
+
+        :return: Iterator[Union[TupleLike, TableLike, None]], producing
+            one TupleLike object, one TableLike object, or None, at a time.
```

Comment:
is this comment accurate?

Reply:
Updated comment to include that an R generator is returned by the UDF and the returned generator is an iterator that yields R Lists.